File: 814-binary-tree-pruning/814-binary-tree-pruning.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<TreeNode,Integer>map;
    boolean del;
    public TreeNode pruneTree(TreeNode root) {
        map=new HashMap<>();
        int sum=root.val+sum(root.right)+sum(root.left);
        if(sum==0){
            return null;
        }
        dfs(root);
        return root;
    }
    void dfs(TreeNode root){
        if(root==null)return;
        if(sum(root.left)==0){
            root.left=null;
        }
        if(sum(root.right)==0){
            root.right=null;
        }
        dfs(root.left);
        dfs(root.right);
    }
    int sum(TreeNode root){
        if(root==null)return 0;
        if(map.containsKey(root))return map.get(root);
        int sum=root.val+sum(root.left)+sum(root.right);
        map.putIfAbsent(root,sum);
        return sum;
    }
    
}


Approach-2 BY @Jay-Thesia

class Solution {
    public TreeNode pruneTree(TreeNode root) {
        
        return helper(root)?null:root;
    }
    
    public boolean helper(TreeNode root){
        
        if(root==null)
            return true;
        
        boolean left=helper(root.left);
        boolean right=helper(root.right);
        
        if(left)
            root.left=null;
        if(right)
            root.right=null;
        
        return root.val==0 && left && right;
    }
}

File: 814_Binary_Tree_Pruning.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int fun(TreeNode* root){
        if(root==NULL){
            return 0;
        }
        if(root->left==NULL && root->right==NULL){
            if(root->val==0){
                return 0;
            }
            else{
                return 1;
            }
        }
        int leftT =fun(root->left);
        int rightT = fun(root->right);
        
        if(leftT==0){
            root->left=NULL;
        }
        if(rightT==0){
            root->right=NULL;
        }
        return leftT+rightT + (root->val==1 ? 1 : 0);
    }
    
    TreeNode* pruneTree(TreeNode* root) {
        if(fun(root)==0){
            return NULL;
        }
        return root;
    }
};
